ls1intum/Artemis
# Move Jenkins build process to Pipelines & make templates configurable.

### Checklist

### Motivation and Context
As discussed in #2169, the current Jenkins build configuration is deprecated:
- Usage of outdated plugins (Multiple SCMs: https://plugins.jenkins.io/multiple-scms/)
- no support for docker builds
- increased setup complexity as all dependencies for all supported programming languages must be installed within the Jenkins master
- Bad scalability: All builds are executed on the Jenkins master and adding more agents is rather expensive as all dependencies must be installed on all agents (and also updated).

This PR moves the Jenkins configuration to so-called Pipelines which define the build process, not through UI settings but declarations in a file.
Exercises that use the old setup should still be supported (e.g. you should still be able to import them into other exercises or start these exercises as student).

### Description
The build process in Jenkins is moved to a `config.xml`-file that shares common steps for all programming languages (e.g. triggering a build when a push to GitLab occurred).
It is extended by a `Jenkinsfile` that is dependent on the used programming language which will be included in the generic `config.xml` file.
The builds steps (including used docker images, the checkout process, the actual build steps, and the reporting of the results to Artemis) is included in the `Jenkinsfile`.

Another improvement is that files from the `src/main/resources/templates/jenkins`-folder can now be modified. This allows (especially non-TUM-institutions) to override settings that might fit well into our deployment, but not so well in different deployments (e.g. a setup where build agents are not static like in our setup, but managed through a Kubernetes cluster).
The `ResourceLoaderService` manages this: Files are first loaded from a `template` folder next to the `artemis.war`-file and if no results are found, the files are loaded from the classpath (e.g. files included in the `.war`.file).
This avoids additional setup steps for all users, but allows experienced users to adjust the jenkins-templates to their own setup.

**Note:** This PR requires at least version `1.1.1` of the Jenkins notification plugin: https://github.com/ls1intum/jenkins-server-notification-plugin. It is already installed at our Jenkins instance at TUM.

Also, some other Jenkins plugins are required for this setup, please refer to the changes in the documentation included in this PR.
The updated documentation also contains steps for setting up remote agents.

#### Possible Questions related to the default `Jenkinsfile`
Q: Why did you use `checkout([$class: 'GitSCM', branches: [[name: '*/master']], doGenerateSubmoduleConfigurations: false, extensions: [], submoduleCfg: [], userRemoteConfigs: [[credentialsId: '#gitCredentials', name: 'tests', url: '#testRepository']]])` instead of `git url: '#testRepository' credentialsId: '#gitCredentials'`?
A: The `git` step might be sufficient for the most cases, but it is recommended to use the `checkout` step instead (see https://www.jenkins.io/doc/pipeline/steps/git/). I decided to use the `checkout` step as the Jenkinsfile might be adjusted by other universities and this shows them more of the available options.

Q: Why must every step of the `Build` stage be within a `timestamps` block?
A: Artemis expects the timestamp to parse the logs.

Q: But why did you not activate it for all pipeline builds? There even is an option to do this.
A: This would also activate the timestamps for builds that are not related to Artemis but might be executed on the Jenkins instance.

Q: Why did you use the `cleanup` action in the `step` to send the results and not the `always` option?
A: There are multiple stages available that can be used to execute commands after the build completed. However, we can not change the order of execution of these steps and `cleanup` is *always* executed last.

Q: Why is there a `|| true` next to `cp target/surefire-reports/*.xml $WORKSPACE/results/`?
A: If the build fails (e.g. due to a compilation error), no test-results are generated (and thus they can not be copied). The `cp` command then exits with a non-zero exit code and Jenkins will then stop the build (and will not sent the test results to Artemis.
To prevent this, we ignore the return value of the `cp` command.

### Steps for Testing
As this affects the Jenkins/GitLab-setup (and not the Bamboo/Bitbucket-setup), please use TS2 to test these changes.
1. Make sure an exercise exists that was created before deploying this PR. Make sure that you have not started it yet.
2. Deploy this PR.
3. Create a new Java or Python exercise (for Java, also try static code analysis) and verify that the solution and base repository are executed correctly (i.e. go to `Edit in Editor` and validate that the solution repository passes all test cases whilst the base repository passes none).
4. Start the exercise as a student and participate: The normal behaviour, including the detection of failed and successful test cases, should work. Also validate that build logs are shown if a build error occured.
5. Import the exercise into a new exercise and validate that the repositories are updated correctly (e.g. by validating that pushing to the new solution repository actually affects the correct exercise)
6. Start the imported exercise as students and repeat the testing from step 4.
7. Repeat steps 4-6 but use the programming exercise that was created before this PR was deployed instead.

Please also test that creating programming exercises still works on the Bamboo/Bitbucket-setup (TS1/TS3).

### Test Coverage

* ExerciseService.java: 85%
* programming-exercise.component.ts 95%
